From 5ce6ffe3d3b545dab65313f593810b692ff8178e Mon Sep 17 00:00:00 2001
From: Jiebing Li <jiebing.li@intel.com>
Date: Thu, 4 Dec 2014 15:43:15 +0800
Subject: [PATCH 218/441] usb/dwc3-device: disable USB reception multi-packet
 thresholding

SYNOPSYS confirms that USB reception multi-packet thresholding
feature doesn't work if DEPCG.BrstSiz is set to 0. So the patch
disables this feature in case gadget function doesn't set
bMaxBurst in ep's companion descriptor (0 by default).

Change-Id: Ib88c1fa9048575c3b4a4d0bf74f842d90fa6f67b
Signed-off-by: Jiebing Li <jiebing.li@intel.com>
(cherry picked from commit e8beb3070bf8ac65e385b654e6003d21b53c6119)

Signed-off-by: Brett T. Warden <brett.t.warden@intel.com>
---
 drivers/usb/dwc3/core.c | 21 +++++++++++++++++++++
 drivers/usb/dwc3/core.h |  4 ++++
 2 files changed, 25 insertions(+)

--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -564,6 +564,25 @@ static void dwc3_suspend_phy(struct dwc3
 	dwc3_writel(dwc->regs, GUSB3PIPECTL0, data);
 }
 
+/*
+ * dwc3_disable_multi_packet - disable reception multi-packet
+ * thresholding in order to support burst size 0 per SYNOPSIS
+ * requirement.
+ */
+static void dwc3_disable_multi_packet(struct dwc3 *dwc)
+{
+	u32			reg;
+
+	reg = dwc3_readl(dwc->regs, DWC3_GRXTHRCFG);
+	if (reg) {
+		reg &= ~DWC3_GRXTHRCFG_USBRXPKTCNTSEL;
+		reg &= ~DWC3_GRXTHRCFG_USBRXPKTCNT_MASK;
+		reg &= ~DWC3_GRXTHRCFG_USBMAXRXBURSTSIZE_MASK;
+
+		dwc3_writel(dwc->regs, DWC3_GRXTHRCFG, reg);
+	}
+}
+
 static int dwc3_handle_otg_notification(struct notifier_block *nb,
 		unsigned long event, void *data)
 {
@@ -964,6 +983,8 @@ static int dwc3_resume_common(struct dev
 
 	spin_lock_irqsave(&dwc->lock, flags);
 
+	dwc3_disable_multi_packet(dwc);
+
 	dwc3_suspend_phy(dwc, false);
 
 	dwc3_writel(dwc->regs, DWC3_GCTL, dwc->gctl);
--- a/drivers/usb/dwc3/core.h
+++ b/drivers/usb/dwc3/core.h
@@ -143,6 +143,10 @@
 /* Bit fields */
 
 /* Global Configuration Register */
+#define DWC3_GRXTHRCFG_USBRXPKTCNTSEL		(1 << 29)
+#define DWC3_GRXTHRCFG_USBRXPKTCNT_MASK		(0xf << 24)
+#define DWC3_GRXTHRCFG_USBMAXRXBURSTSIZE_MASK	(0x1f << 19)
+
 #define DWC3_GCTL_PWRDNSCALE(n)	((n) << 19)
 #define DWC3_GCTL_U2RSTECN	(1 << 16)
 #define DWC3_GCTL_RAMCLKSEL(x)	(((x) & DWC3_GCTL_CLK_MASK) << 6)
